src/OSPSuite.Core/Domain/Services/ResidualCalculator.cs

```diff
@@ -99,20 +99,45 @@ private IReadOnlyList<Residual> calculateOutputResiduals(DataColumn simulationCo
          {
             var weight = outputMapping.Weight * outputMapping.WeightedObservedData.Weights[index];
             var observedValue = convertToBaseUnit(mergedDimension, currentObservedDataUnit, observedValueColumn[index]);
-            if (!observedValue.IsValid())
-               continue;
 
             var observedTime = observedTimeColumn[index];
-            var simulatedValue = simulationColumn.GetValue(observedTime);
-            if (!simulatedValue.IsValid())
+
+            if(simulationTimeIsTooShort(simulationColumn, observedTime))
+               continue;
+
+            if(observationShouldBeIgnored(observedValue))
                continue;
 
-            outputResiduals.Add(new Residual(observedTime, weight * residualCalculatorFunc(simulatedValue, observedValue, lloq), weight));
+            var simulatedValue = simulationColumn.GetValue(observedTime);
+
+            outputResiduals.Add(!simulatedValue.IsValid() ? 
+               new Residual(observedTime, residualValueForInvalidSimulationValue(simulatedValue), weight) : 
+               new Residual(observedTime, weight * residualCalculatorFunc(simulatedValue, observedValue, lloq), weight));
          }
 
          return outputResiduals;
       }
 
+      private static bool observationShouldBeIgnored(float observedValue)
+      {
+         return !observedValue.IsValid();
+      }
+
+      private static bool simulationTimeIsTooShort(DataColumn simulationColumn, float observedTime)
+      {
+         return simulationColumn.BaseGrid.Values.Max() < observedTime;
```

Comment:
Another small optimization: `simulationColumn.BaseGrid.Values.Max()` could be evaluated and cached only once before the `foreach (var index in observedTimeIndices)` loop